### PR TITLE
Uml++ with relatedTo and relatedFrom

### PR DESCRIPTION
**Pull request by hchoura**

### CODE REVIEW

File: src/foam/doc/DocBrowser.js [hunk 1]
```diff
@@ -532,3 +559,954 @@ foam.debug.doc = function(opt_obj, showUnused) {
...
+
+foam.CLASS({
+  package: 'foam.doc',
+  name: 'UMLDiagram',
+  extends: 'foam.u2.Element',
+
+  implements: ['foam.memento.MementoMgr'],
+  imports: ['browserPath'],
+  requires: [
+    'foam.doc.DocBorder',
```
Comment (kgrgreer) [inline src/foam/doc/DocBrowser.js]: Sort requires list.

File: src/foam/doc/DocBrowser.js [hunk 2]
```diff
@@ -532,3 +559,954 @@ foam.debug.doc = function(opt_obj, showUnused) {
...
+  ],
+
+  exports: ['as data'],
+
+  constants: {
+    SELECTED_COLOR: 'white', //#ddd',
+    UNSELECTED_COLOR: '#FFFFCC'
+  },
+
+  axioms: [
```
Comment (kgrgreer) [inline src/foam/doc/DocBrowser.js]: We've recently added a 'css' property to Model so you can more easily do this as just:
css: `
  css here
  and here
  ...
`,

File: src/foam/doc/DocBrowser.js [hunk 3]
```diff
@@ -532,3 +559,954 @@ foam.debug.doc = function(opt_obj, showUnused) {
...
+      start(this.canvas).
+      on('click', this.onClick).
+      end().
+      end();
+    },
+
+    function sign(ex, sx) {
+      if (ex - sx > 0) {
+        return 1;
+      } else if (ex - sx < 0) {
```
Comment (kgrgreer) [inline src/foam/doc/DocBrowser.js]: No need for else's if you've already returned.

File: src/foam/doc/DocBrowser.js [hunk 4]
```diff
@@ -532,3 +559,954 @@ foam.debug.doc = function(opt_obj, showUnused) {
...
+        return -1;
+      } else {
+        return 0;
+      }
+    },
+
+    function dashedLine(d, sx, sy, ex, ey) {
+      var dashedLine, changeX = 0,
```
Comment (kgrgreer) [inline src/foam/doc/DocBrowser.js]: No need to assign to 0 if you're just reassigning on next two lines. Just set to sign() values when declaring.

File: src/foam/doc/DocBrowser.js [hunk 5]
```diff
@@ -532,3 +559,954 @@ foam.debug.doc = function(opt_obj, showUnused) {
...
+        }
+      }
+    },
+
+    function addRequires(x, y, w, h) {
+      var triangleSize = 5;
+      var marge = 5;
+      var d = 300;
+      var cls = this.data;
+      if (cls.model_.requires !== undefined) {
```
Comment (kgrgreer) [inline src/foam/doc/DocBrowser.js]: Here and elsewhere, see the FOAM style guideline regarding spacing: https://github.com/foam-framework/foam2/blob/master/doc/guides/StyleGuide.md

File: src/foam/doc/DocBrowser.js [hunk 6]
```diff
@@ -532,3 +559,960 @@ foam.debug.doc = function(opt_obj, showUnused) {
...
+  ],
+
+  exports: [ 'as data' ],
+
+  constants: {
+    SELECTED_COLOR: 'white', //#ddd',
+    UNSELECTED_COLOR: '#FFFFCC'
+  },
+
+  css: [
```
Comment (kgrgreer) [inline src/foam/doc/DocBrowser.js]: css: `
                ^ { width:1200px; margin: 20px; }
                ^ canvas { border: 1px solid black; }
                ^ .foam-u2-ActionView- { margin: 10px; }
                ^ input[type='range'] { width: 400px; }
`

File: src/foam/doc/DocBrowser.js [hunk 7]
```diff
@@ -532,3 +559,960 @@ foam.debug.doc = function(opt_obj, showUnused) {
...
+      this.addSubClasses(this.canvas.width / 2 - widthCenterBox / 2, this.canvas.height / 2 - heightCenterBox * 2);
+      /*this.addImports(this.canvas.width / 2 - widthCenterBox / 2, this.canvas.height / 2 - heightCenterBox * 2);
+      this.addExports(this.canvas.width / 2 - widthCenterBox / 2, this.canvas.height / 2 - heightCenterBox * 2);*/
+      this.addRelatedto(this.canvas.width / 2 - widthCenterBox / 2, this.canvas.height / 2 - heightCenterBox * 2);
+      this.addRelatedFrom(this.canvas.width / 2 - widthCenterBox / 2, this.canvas.height / 2 - heightCenterBox * 2);
+      this.addLegend();
+
+      this.
+      addClass(this.myClass()).
+      start('center').
```
Comment (kgrgreer) [inline src/foam/doc/DocBrowser.js]: Indent two spaces after the start() until the end. Like:
start().
  tag('br')
  ...
end()

File: src/foam/doc/DocBrowser.js [hunk 8]
```diff
@@ -532,3 +559,960 @@ foam.debug.doc = function(opt_obj, showUnused) {
...
+
+    function sign( ex, sx ) {
+      if ( ex - sx > 0 ) {
+        return 1;
+      } if ( ex - sx < 0 )
+          return -1;
+        return 0;
+    },
+
+    function dashedLine( sx, sy, ex, ey ) {
```
Comment (kgrgreer) [inline src/foam/doc/DocBrowser.js]: Faster and easier to draw dashed lines this way:
https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/setLineDash

File: src/foam/doc/DocBrowser.js [hunk 9]
```diff
@@ -584,19 +584,7 @@ foam.CLASS({
...
-        /*
-               ^ { width:1200px; margin: 20px; }
-               ^ canvas { border: 1px solid black; }
-               ^ .foam-u2-ActionView- { margin: 10px; }
-               ^ input[type='range'] { width: 400px; }
-               */
-      }
-    })
-  ],
+  css: ^ { width:1200px; margin: 20px; } ^ canvas { border: 1px solid black; } ^ .foam-u2-ActionView- { margin: 10px; } ^ input[type='range'] { width: 400px; },
```
Comment (kgrgreer) [inline src/foam/doc/DocBrowser.js]: This needs to be quoted, preferably in a multi-line quote (`). Doesn't this cause a syntax error?

Comment (hchoura) [reply src/foam/doc/DocBrowser.js]: No, it does not.

File: src/foam/doc/DocBrowser.js [hunk 10]
```diff
@@ -532,3 +559,947 @@ foam.debug.doc = function(opt_obj, showUnused) {
...
+  ],
+
+  exports: [ 'as data' ],
+
+  constants: {
+    SELECTED_COLOR: 'white', //#ddd',
+    UNSELECTED_COLOR: '#FFFFCC'
+  },
+
+  css: ^ { width:1200px; margin: 20px; } ^ canvas { border: 1px solid black; } ^ .foam-u2-ActionView- { margin: 10px; } ^ input[type='range'] { width: 400px; },
```
Comment (kgrgreer) [inline src/foam/doc/DocBrowser.js]: Is not valid code. Will produce syntax error.

**Pull request merged**